crates/matrix-sdk/src/room/timeline/event_handler.rs

```diff
@@ -0,0 +1,460 @@
+    Remote { event_id: OwnedEventId, txn_id: Option<OwnedTransactionId>, position: EventPosition },
+}
+
+impl Flow {
+    fn to_key(&self) -> TimelineKey {
+        match self {
+            Self::Remote { event_id, .. } => TimelineKey::EventId(event_id.to_owned()),
+            Self::Local { txn_id } => TimelineKey::TransactionId(txn_id.to_owned()),
+        }
+    }
+}
+
+struct EventMetadata {
+    raw_event: Option<Raw<AnySyncTimelineEvent>>,
+    sender: OwnedUserId,
+    origin_server_ts: Option<MilliSecondsSinceUnixEpoch>,
+    is_own_event: bool,
+    relations: Option<Relations>,
+    encryption_info: Option<EncryptionInfo>,
+}
+
+impl From<AnySyncTimelineEvent> for EventKind {
+    fn from(event: AnySyncTimelineEvent) -> Self {
+        match event {
+            AnySyncTimelineEvent::MessageLike(AnySyncMessageLikeEvent::RoomRedaction(
+                SyncRoomRedactionEvent::Original(OriginalSyncRoomRedactionEvent {
+                    redacts,
+                    content,
+                    ..
+                }),
+            )) => Self::Redaction { redacts, content },
+            AnySyncTimelineEvent::MessageLike(ev) => match ev.original_content() {
+                Some(content) => Self::Message { content },
+                None => Self::RedactedMessage,
+            },
+            AnySyncTimelineEvent::State(ev) => match ev.original_content() {
+                Some(_content) => Self::State { _content },
+                None => Self::RedactedState,
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+enum EventKind {
+    Message { content: AnyMessageLikeEventContent },
+    RedactedMessage,
+    Redaction { redacts: OwnedEventId, content: RoomRedactionEventContent },
+    // FIXME: Split further for state keys of different type
+    State { _content: AnyStateEventContent },
+    RedactedState, // AnyRedactedStateEventContent
+}
+
+enum EventPosition {
+    Start,
+    End,
+}
+
+struct EventHandler<'a> {
+    meta: EventMetadata,
+    flow: Flow,
+    timeline: &'a TimelineInner,
+    event_added: bool,
+}
+
+impl<'a> EventHandler<'a> {
+    fn new(meta: EventMetadata, flow: Flow, timeline: &'a TimelineInner) -> Self {
+        Self { meta, flow, timeline, event_added: false }
+    }
+
+    fn handle_event(mut self, event_kind: EventKind) {
+        match event_kind {
+            EventKind::Message { content } => match content {
+                AnyMessageLikeEventContent::Reaction(c) => self.handle_reaction(c),
+                AnyMessageLikeEventContent::RoomMessage(c) => self.handle_room_message(c),
+                // TODO
+                _ => {}
+            },
+            EventKind::RedactedMessage => {
+                self.add(NewEventTimelineItem::redacted_message());
+            }
+            EventKind::Redaction { redacts, content } => self.handle_redaction(redacts, content),
+            // TODO: State events
+            _ => {}
+        }
+
+        if !self.event_added {
```

Comment:
Can't this be handled like the `did_update` flag? Where we return the value instead of mutating things in some magic place somewhere else?

Reply:
Well, there is only a single place where it is set, so I think it's still obvious enough what happens and this way the method signatures are shorter. We can try out the other way though.